comms/tests/control_service/service.rs

```diff
@@ -0,0 +1,170 @@
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::support::utils;
+use serde::{Deserialize, Serialize};
+
+use std::{convert::TryInto, sync::RwLock, thread, time::Duration};
+
+use tari_comms::{
+    connection::{curve_keypair, Connection, Context, Direction, Linger},
+    control_service::{
+        messages::EstablishConnection,
+        ControlService,
+        ControlServiceConfig,
+        ControlServiceError,
+        ControlServiceMessageContext,
+    },
+    dispatcher::{DispatchError, DispatchResolver, Dispatcher},
+    message::{Message, MessageEnvelope, MessageError, MessageFlags, MessageHeader, NodeDestination},
+    types::MessageEnvelopeHeader,
+};
+use tari_crypto::{keys::PublicKey, ristretto::RistrettoPublicKey};
+use tari_utilities::message_format::MessageFormat;
+
+#[derive(Eq, PartialEq, Hash, Serialize, Deserialize)]
+enum MessageType {
+    EstablishConnection = 1,
+    InvalidMessage = 2,
+}
+
+lazy_static! {
+    static ref HANDLER_CALL_COUNT: RwLock<u8> = RwLock::new(0);
+}
+
+fn test_handler(context: ControlServiceMessageContext) -> Result<(), ControlServiceError> {
+    let msg: EstablishConnection = context
+        .message
+        .to_message()
+        .map_err(|err| DispatchError::HandlerError(format!("Failed to parse message: {}", err)))?;
+
+    assert!(msg.address.is_ip());
+
+    let mut lock = HANDLER_CALL_COUNT.write().unwrap();
+    *lock += 1;
+
+    Ok(())
+}
+
+struct CustomTestResolver;
+
+impl DispatchResolver<MessageType, ControlServiceMessageContext> for CustomTestResolver {
+    fn resolve(&self, context: &ControlServiceMessageContext) -> Result<MessageType, DispatchError> {
+        let header: MessageHeader<MessageType> = context
+            .message
+            .to_header()
+            .map_err(|err| DispatchError::HandlerError(format!("Failed to parse header: {}", err)))?;
+
+        Ok(header.message_type)
+    }
+}
+
+fn construct_envelope<T: MessageFormat>(
+    pk: RistrettoPublicKey,
+    message_type: MessageType,
+    msg: T,
+) -> Result<MessageEnvelope, MessageError>
+{
+    let msg_header = MessageHeader { message_type };
+    let msg: Message = (msg_header, msg).try_into()?;
+    let envelope_header = MessageEnvelopeHeader {
+        source: pk,
+        dest: NodeDestination::Unknown,
+        flags: MessageFlags::empty(),
+        signature: vec![0],
+        version: 0,
+    };
+
+    Ok(MessageEnvelope::new(
+        vec![0],
+        envelope_header.to_binary()?,
+        msg.to_binary()?,
+    ))
+}
+
+fn poll_handler_call_count_change(initial: u8, ms: u64) -> Option<u8> {
+    for _i in 0..ms {
+        {
+            let lock = HANDLER_CALL_COUNT.read().unwrap();
+            if *lock != initial {
+                return Some(*lock);
+            }
+        }
+        thread::sleep(Duration::from_millis(1))
+    }
+
+    None
+}
+
+#[test]
+fn recv_message() {
+    // tari_common::logging::initialize_logger::initialize_logger();
```

Comment:
You don't need this. It's intended for applications. Just call debug! as usual and run with `-- --nocapture` on the CLI to see debug output

Reply:
Hmm, I know about `--nocapture` but that never worked for me with the log macros. I think that the `log` crate doesn't have a stdout consumer by default. Probably pretty simple to set that up.